content/blog/refinements-in-ruby.md

```diff
@@ -0,0 +1,120 @@
+image = "/images/2024/refinements-in-ruby/header.jpeg"
+draft = false
+authors = ["Syed Mohd Mehndi"]
+description = "Refinements in Ruby: A Flexible Approach to Modifying Core Classes"
+tags = ["Ruby"]
+categories = ["Ruby"]
+type = ""
++++
+
+Ruby is a highly flexible programming language known for its dynamic capabilities and metaprogramming power. One such powerful feature in Ruby is refinements. Introduced in Ruby 2.0, refinements allow you to modify existing classes (even core classes like String, Array, etc.) in a more controlled and localized way. This is particularly useful when you want to change class behavior in specific contexts without affecting the global environment, which is a common issue with traditional monkey-patching.
+
+In this blog, we’ll dive into the basics of using refinements, explore common use cases, and explain why they can be preferable to monkey-patching.
+
+### What Are Refinements?
+
+Refinements offer a way to extend or override methods of existing classes, but only within a specific scope (like a module or class). This helps prevent unintended side effects that can occur when modifying core classes or third-party libraries globally. In other words, refinements give you a safer alternative to monkey-patching by limiting the scope of changes.
+
+### Key Characteristics of Refinements:
+
+Scoped Modifications: Refinements apply only in specific contexts (like within a module or class).
+Opt-in Behavior: You have to explicitly “activate” a refinement within the scope where you want it to apply.
+No Global Impact: Unlike monkey-patching, which modifies behavior globally, refinements are local to the scope they are activated in.
+
+### Refinements in Action
+
+Let’s walk through an example to demonstrate how refinements work in Ruby.
+
+#### Step 1: Defining a Refinement
+
+Here’s how you define a refinement in Ruby. We will modify the String class to add a new method, #reverse_words, which reverses the words in a string, not the characters.
+
+```ruby
+module StringExtensions
+    refine String do
+        def reverse_words
+          self.split(' ').reverse.join(' ')
+        end
+    end
+end
+```
+
+In this example, we created a module StringExtensions that contains a refinement of the String class. This refinement adds a method reverse_words to reverse the order of words in a string.
+
+#### Step 2: Activating the Refinement
+
+To use the refinement, we need to activate it in a specific scope using the using keyword:
+
+```ruby
+class SentenceManipulator
+    using StringExtensions
+
+    def initialize(sentence)
+      @sentence = sentence
+    end
+
+    def reverse
+      @sentence.reverse_words
+    end
+end
+
+sentence = SentenceManipulator.new("Hello World from Ruby")
+puts sentence.reverse
+# Output: "Ruby from World Hello"
+```
+
+Here, the refinement applies only inside the SentenceManipulator class. If you try to use reverse_words outside of this class, it will not be available.
```

Comment:
```suggestion
Here, the refinement applies only inside the `SentenceManipulator` class. If you try to use reverse_words outside of this class, it will not be available.
```

Reply:
Done, Thanks